### PR TITLE
Add improved find

### LINKED ISSUE #90: Improve 'find' command
As a user, I can search the addressbook by names, tags and any other information so that I can find the patients and appointments that match all of my keywords faster.

### PR DESCRIPTION
**Pull request by NicholasTYD**
Closes #90 

Change find command to the following format:
`find [n/NAME] [p/PHONE] [e/EMAIL] [a/ADDRESS] [t/TAG]…​ [r/REASON] [ds/DATE_START] [de/DATE_END]`

Todo:
- [x] Add improved find
- [x] Fix broken test cases
- [x] Add new test cases
- [x] Edit user guide for improved find
- [x] Fix checkstyle and javadocs

### CODE REVIEW

**Review by JovonLim — APPROVED** [review]
LGTM overall, but is it necessary to have so many predicates when they have no differences?

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 1]
```diff
@@ -19,15 +36,81 @@ public class FindCommandParser implements Parser<FindCommand> {
...
+
+    private void checkIfInputsValid(String name, String phone, String email, String address, List<String> tagList,
+                                    String reason, String startDateTime, String endDateTime) throws ParseException {
+
+        boolean areAllFieldsEmpty = name.isEmpty() && phone.isEmpty() && email.isEmpty()
+                && address.isEmpty() && tagList.isEmpty() && reason.isEmpty() && startDateTime.isEmpty()
+                && endDateTime.isEmpty();
+        if (areAllFieldsEmpty) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
         }
```
[the remark this answers is not in the document]
Comment (NicholasTYD) [reply src/main/java/seedu/address/logic/parser/FindCommandParser.java]: Sure

**Comment by NicholasTYD** [comment]
> LGTM overall, but is it necessary to have so many predicates when they have no differences? Using enum might be better in reducing code repetition

I think that it may be easier to spilt up the predicates as it may be easier during testing to troubleshoot errors for individual fields if bugs are encountered.

**Review by Guanzhou03 — COMMENTED** [review]
LGTM, but as Jovon mentioned, perhaps code duplication can be reduced with the use of an enumeration. Here's an example that I am working on which is a single class that implements Predicate<Appointment>, and will be able handle hiding appointments by tag, keywords, and isMarked.

Screenshot 2022-10-16 at 2 03 58 PM

**Review by Guanzhou03 — APPROVED** [review]

**Pull request merged**